pkg/pool-stable/contracts/meta/MetaStablePool.sol

```diff
@@ -0,0 +1,372 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-pool-utils/contracts/interfaces/IPriceOracle.sol";
+import "@balancer-labs/v2-pool-utils/contracts/oracle/PoolPriceOracle.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/BalancerErrors.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/LogCompression.sol";
+
+import "../StablePool.sol";
+import "./OracleMiscData.sol";
+import "./StableOracleMath.sol";
+
+contract MetaStablePool is StablePool, StableOracleMath, PoolPriceOracle, IPriceOracle {
+    using OracleMiscData for bytes32;
+
+    event OracleEnabledChanged(bool enabled);
+
+    constructor(
+        IVault vault,
+        string memory name,
+        string memory symbol,
+        IERC20[] memory tokens,
+        uint256 amplificationParameter,
+        uint256 swapFeePercentage,
+        uint256 pauseWindowDuration,
+        uint256 bufferPeriodDuration,
+        bool oracleEnabled,
+        address owner
+    )
+        StablePool(
+            vault,
+            name,
+            symbol,
+            tokens,
+            amplificationParameter,
+            swapFeePercentage,
+            pauseWindowDuration,
+            bufferPeriodDuration,
+            owner
+        )
+    {
+        _require(tokens.length == 2, Errors.NOT_TWO_TOKENS);
+        _setOracleEnabled(oracleEnabled);
+    }
+
+    // Swap
+
+    /**
+     * Update price oracle with the pre-swap balances
+     * Note this function does not perform any safety checks about swaps, it relies on upper implementations for that.
+     */
+    function _onSwapGivenIn(
+        SwapRequest memory request,
+        uint256[] memory balances,
+        uint256 indexIn,
+        uint256 indexOut
+    ) internal virtual override whenNotPaused returns (uint256) {
+        _updateOracle(request.lastChangeBlock, balances[0], balances[1]);
```

Comment:
The base classes don't perform any safety checks, in particular, they don't check that `msg.sender == vault` (as WeightedPool2Tokens does). We should add this check either here or in BaseMinimalSwapInfoPool and BaseGeneralPool (I think this might be safer).

Reply:
Let's add it here for now, as it does not apply necessarily for the current stable and weighted pools